pkg/controller/eventlog/eventlog.go

```diff
@@ -0,0 +1,52 @@
+package eventlog
+
+import (
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type ControllerWithEvents interface {
+	EventRecorder() *record.EventRecorder
+	Logger() *logr.Logger
+	Client() *client.Client
+}
+
+type ControllerWithLogger interface {
+	Name() string
+	Logger() *logr.Logger
+}
+
+type ErrorEvent struct {
+	E      error
+	Reason string
+	Owner  *runtime.Object
+}
+
+func (eev *ErrorEvent) Error() string {
+	return eev.E.Error()
+}
+
+func ValuedLogger(cwl ControllerWithLogger, keysAndValues ...interface{}) logr.Logger {
+	log := Logger(cwl)
+	logger := log.WithValues(keysAndValues...)
+	return logger
+}
+
+func Logger(cwl ControllerWithLogger) logr.Logger {
+	if *cwl.Logger() != nil {
+		return *cwl.Logger()
+	}
+	*cwl.Logger() = logf.Log.WithName(cwl.Name())
+	return *cwl.Logger()
+}
+
+func LogAndSendEvent(cwe ControllerWithEvents, owner runtime.Object, reason, message string) {
```

Comment:
@rigazilla what do you think to introduce ReasonType to avoid misprints between reason and message positions?

Reply:
@dmvolod sounds good

Reply:
@dmvolod,  mmm my idea was to add something like
```
type ReasonType string
```
but it doesn't enforce that much cause string is implicitly converted to ReasonType.
Any better easy idea? Or I would leave things like this

Reply:
@rigazilla you should define const as `ReasonType` and seems to you don't need to convert except last call to the event API.
Probably I'm wrong, but should work

Reply:
actually it works, but it doesn't enforce anything on the content of the reason parameter, since any  free string is implicitly converted to ReasonType without any warning.
I've just realized that also the k8s.Event uses string for both reason and message args, I would do same